src/include/storage/storage_structure/overflow_file.h

```diff
@@ -0,0 +1,159 @@
+        pageCache.clear();
+        this->nextPosToWriteTo = nextPosToWriteTo;
+    }
+
+private:
+    uint8_t* addANewPage();
+    void setStringOverflow(
+        const char* inMemSrcStr, uint64_t len, common::ku_string_t& diskDstString);
+
+    inline void read(transaction::TransactionType trxType, common::page_idx_t pageIdx,
+        const std::function<void(uint8_t*)>& func) const;
+
+private:
+    static const common::page_idx_t END_OF_PAGE =
+        common::BufferPoolConstants::PAGE_4KB_SIZE - sizeof(common::page_idx_t);
+    // This is the index of the last free byte to which we can write.
+    PageCursor& nextPosToWriteTo;
+    OverflowFile& overflowFile;
+    // Cached pages which have been written in the current transaction
+    std::unordered_map<common::page_idx_t, std::unique_ptr<InMemPage>> pageCache;
+};
+
+// Stores the current state of the overflow file
+// The number of pages in use are stored here so that we can write new pages directly
+// to the overflow file, and in the case of an interruption and rollback the header will
+// still record the correct place in the file to allocate new pages
+struct StringOverflowFileHeader {
+    common::page_idx_t pages;
+    PageCursor cursors[NUM_HASH_INDEXES];
+
+    StringOverflowFileHeader() : pages{1} {
+        std::fill(cursors, cursors + NUM_HASH_INDEXES, PageCursor());
+    }
+};
+
+class OverflowFile {
+public:
+    // For reading an existing overflow file
+    OverflowFile(const DBFileIDAndName& dbFileIdAndName, BufferManager* bufferManager, WAL* wal,
+        bool readOnly, common::VirtualFileSystem* vfs);
+
+    // For creating an overflow file from scratch
+    OverflowFile(const std::string& fName, common::VirtualFileSystem* vfs)
+        : numPagesOnDisk{0}, fileHandle{std::make_unique<FileHandle>(
+                                 fName, FileHandle::O_PERSISTENT_FILE_CREATE_NOT_EXISTS, vfs)},
+          bufferManager{nullptr}, wal{nullptr}, pageCounter{0}, headerChanged{true} {
+        // Reserve a page for the header
+        getNewPageIdx();
+    }
+
+    // Handles contain a reference to the overflow file
+    OverflowFile(OverflowFile&& other) = delete;
+
+    void rollbackInMemory();
+    void prepareCommit();
+    void checkpointInMemory();
+
+    static inline DBFileIDAndName constructDBFileIDAndName(
+        const DBFileIDAndName& dbFileIdAndNameForMainDBFile) {
+        DBFileIDAndName copy = dbFileIdAndNameForMainDBFile;
+        copy.dbFileID.isOverflow = true;
+        copy.fName = StorageUtils::getOverflowFileName(dbFileIdAndNameForMainDBFile.fName);
+        return copy;
+    }
+
+    inline OverflowFileHandle* addHandle() {
+        KU_ASSERT(handles.size() < NUM_HASH_INDEXES);
+        handles.emplace_back(
+            std::make_unique<OverflowFileHandle>(*this, header.cursors[handles.size()]));
+        return handles.back().get();
+    }
+
+    inline common::page_idx_t getNewPageIdx() {
```

Comment:
Ditto here on public.

Reply:
For this one, and `setHeaderChanged`, they are used by the handles. I guess the handles could be made friend classes if we want to have a more restricted interface for the hash index to use.